File: common/src/androidMain/kotlin/com/blockstream/common/devices/NfcCardManager.java
//package com.blockstream.common.devices;
//
//import android.nfc.NfcAdapter;
//import android.nfc.Tag;
//import android.nfc.tech.IsoDep;
//import android.os.SystemClock;
//import android.util.Log;
//
//import java.io.IOException;
//
///**
// * Manages connection of NFC-based cards. Extends Thread and must be started using the start() method. The thread has
// * a runloop which monitors the connection and from which CardListener callbacks are called.
// */
//public class NfcCardManager extends Thread implements NfcAdapter.ReaderCallback {
//    private static final String TAG = "NFCCardManager";
//    private static final int DEFAULT_LOOP_SLEEP_MS = 50;
//
//    private IsoDep isoDep;
//    private boolean isRunning;
//    private CardListener cardListener;
//    private int loopSleepMS;
//
////  static {
////    Crypto.addBouncyCastleProvider();
////  }
//
//    /**
//     * Constructs an NFC Card Manager with default delay between loop iterations.
//     */
//    public NfcCardManager() {
//        this(DEFAULT_LOOP_SLEEP_MS);
//    }
//
//    /**
//     * Constructs an NFC Card Manager with the given delay between loop iterations.
//     *
//     * @param loopSleepMS time to sleep between loops
//     */
//    public NfcCardManager(int loopSleepMS) {
//        this.loopSleepMS = loopSleepMS;
//    }
//
//    /**
//     * True if connected, false otherwise.
//     * @return if connected, false otherwise
//     */
//    public boolean isConnected() {
//        try {
//            return isoDep != null && isoDep.isConnected();
//        } catch (Exception e) {
//            e.printStackTrace();
//            return false;
//        }
//    }
//
//    @Override
//    public void onTagDiscovered(Tag tag) {
//        isoDep = IsoDep.get(tag);
//        try {
//            isoDep = IsoDep.get(tag);
//            isoDep.connect();
//            isoDep.setTimeout(120000);
//        } catch (IOException e) {
//            Log.e(TAG, "error connecting to tag");
//        }
//    }
//
//    /**
//     * Runloop. Do NOT invoke directly. Use start() instead.
//     */
//    public void run() {
//        boolean connected = isConnected();
//
//        while (true) {
//            boolean newConnected = isConnected();
//            if (newConnected != connected) {
//                connected = newConnected;
//                Log.i(TAG, "tag " + (connected ? "connected" : "disconnected"));
//
//                if (connected && !isRunning) {
//                    onCardConnected();
//                } else {
//                    onCardDisconnected();
//                }
//            }
//
//            SystemClock.sleep(loopSleepMS);
//        }
//    }
//
//    /**
//     * Reacts on card connected by calling the callback of the registered listener.
//     */
//    private void onCardConnected() {
//        isRunning = true;
//
//        if (cardListener != null) {
//            cardListener.onConnected(new NfcCardChannel(isoDep));
//        }
//
//        isRunning = false;
//    }
//
//    /**
//     * Reacts on card disconnected by calling the callback of the registered listener.
//     */
//    private void onCardDisconnected() {
//        isRunning = false;
//        isoDep = null;
//        if (cardListener != null) {
//            cardListener.onDisconnected();
//        }
//    }
//
//    /**
//     * Sets the card listener.
//     *
//     * @param listener the new listener
//     */
//    public void setCardListener(CardListener listener) {
//        cardListener = listener;
//    }
//}

File: common/src/androidMain/kotlin/com/blockstream/common/devices/NfcCardManager.kt
package com.blockstream.common.devices

import android.nfc.NfcAdapter
import android.nfc.Tag
import android.nfc.tech.IsoDep
import android.os.SystemClock
import android.util.Log
import kotlinx.io.IOException

/**
 * Manages connection of NFC-based cards. Extends Thread and must be started using the start() method. The thread has
 * a runloop which monitors the connection and from which CardListener callbacks are called.
 */
class NfcCardManager @JvmOverloads constructor(
    private val loopSleepMS: Long = DEFAULT_LOOP_SLEEP_MS
) : Thread(), NfcAdapter.ReaderCallback {

    private var isoDep: IsoDep? = null
    private var isRunning = false
    private var cardListener: CardListener? = null

    /**
     * True if connected, false otherwise.
     * @return if connected, false otherwise
     */
    val isConnected: Boolean
        get() = try {
            isoDep?.isConnected == true
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }

    override fun onTagDiscovered(tag: Tag) {
        isoDep = IsoDep.get(tag)?.apply {
            try {
                connect()
                timeout = 120000
            } catch (e: IOException) {
                Log.e(TAG, "error connecting to tag")
            }
        }
    }

    /**
     * Runloop. Do NOT invoke directly. Use start() instead.
     */
    override fun run() {
        var connected = isConnected

        while (true) {
            val newConnected = isConnected
            if (newConnected != connected) {
                connected = newConnected
                Log.i(TAG, "tag ${if (connected) "connected" else "disconnected"}")

                if (connected && !isRunning) {
                    onCardConnected()
                } else {
                    onCardDisconnected()
                }
            }

            SystemClock.sleep(loopSleepMS)
        }
    }

    /**
     * Reacts on card connected by calling the callback of the registered listener.
     */
    private fun onCardConnected() {
        isRunning = true

        cardListener?.onConnected(NfcCardChannel(isoDep!!))

        isRunning = false
    }

    /**
     * Reacts on card disconnected by calling the callback of the registered listener.
     */
    private fun onCardDisconnected() {
        isRunning = false
        isoDep = null
        cardListener?.onDisconnected()
    }

    /**
     * Sets the card listener.
     *
     * @param listener the new listener
     */
    fun setCardListener(listener: CardListener) {
        cardListener = listener
    }

    companion object {
        private const val TAG = "NFCCardManager"
        private const val DEFAULT_LOOP_SLEEP_MS: Long = 50
    }
}